search-app/src/cmr/search/api/concepts_search.clj

```diff
@@ -70,6 +80,74 @@
       :too-many-requests
       "Excessive query rate. Please contact support@earthdata.nasa.gov.")))
 
+(defn- reject-all-granule-query?
+  "Returns true if the all granule query will be rejected."
+  [headers]
+  (and (false? (allow-all-granule-params-flag))
+       (or (not (some? (get headers "client-id")))
+           (not (= "true" (get headers (allow-all-gran-header)))))))
+
+(defn- handle-all-granule-params
+  "Throws error if all granule params needs to be rejected."
+  [headers]
+  (let [err-msg (str "The CMR does not currently allow querying across granules in all "
+                     "collections. To help optimize your search, you should limit your "
+                     "query using conditions that identify one or more collections, "
+                     "such as provider, provider_id, concept_id, collection_concept_id, "
+                     "short_name, version or entry_title. Visit the CMR Client Developer "
+                     "Forum at https://wiki.earthdata.nasa.gov/display/CMR/"
+                     "Granule+Queries+Now+Require+Collection+Identifiers for more "
+                     "information, and for any questions please contact "
+                     "support@earthdata.nasa.gov.")]  
+    (when (reject-all-granule-query? headers)
+      (svc-errors/throw-service-error :bad-request err-msg))))
+
+(defn- illegal-concept-id-in-granule-query?
+  "Check to see if any concept ids are not starting with G and C."
+  [concept-id-param]
+  (when (some? concept-id-param)
+    (let [concept-id-param-list (if (sequential? concept-id-param)
+                                  (remove #(= "" %) concept-id-param)
+                                  [concept-id-param])]
+      (some #(not (re-find #"^[GC]" %)) concept-id-param-list))))
+
+(defn empty-string-values? 
+  "This function is used by remove-map-keys function.
+   The keys will be removed if their values are empty strings, or if it's sequential and
+   only contain empty strings.
+   Note:  When you pass a parameter without = sign, it is considered a nil value and get  
+   filtered out before getting to the params; with the = sign, no value, it becomes empty 
+   string."
+  [v]
+  (or (= "" v)
+      (when (sequential? v)
+        (= [] (remove #(= "" %) v)))))
+
+(defn- all-granule-params?
+  "Returns true if it's a all granule query params.
```

Comment:
Nitpick: "Return true if it's an all-granule query"

Reply:
Done.